src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusExporterMiddleware.cs

```diff
@@ -67,34 +67,40 @@ public async Task InvokeAsync(HttpContext httpContext)
                 return;
             }
 
-            try
+            this.exporter.OnExport = (metrics) =>
             {
-                this.exporter.Collect(Timeout.Infinite);
+                try
+                {
+                    WriteMetricsToResponse(this.exporter, metrics, response).ConfigureAwait(false).GetAwaiter().GetResult();
```

Comment:
(Nit: Don't really need the `.ConfigureAwait(false)` on this line or the one above in the HttpServer version.)

The only thing I don't love about this design (or #2511) and what I was trying to fix with #2517 is this line right here. Ideally we don't call `.Wait()`, `.Result`, or `GetAwaiter().GetResult()` because that is the "sync-over-async" anti-pattern. Ties up two threads and can lead to starvation. Would it be the end of the world to do? Probably not, because we have the semaphore above, but I think it is a code smell that we don't have an async API where we need one.

Reply:
Based on the offline discussion, I've updated the PR, there are still many places I haven't covered (to avoid the PR getting too big) https://github.com/open-telemetry/opentelemetry-dotnet/pull/2553/files#r744332289.